### ISSUE TITLE
Correct the lossRates and productionRates output

### ISSUE THREAD

**Issue by spco** [AtChem/AtChem2#319]
`solverFunctions:66-74` calculate lossRates and productionRates wrongly. Proposed solution:

Switch to use only one config file to replace `outputLossRates.config` and `outputProductionRates.config`. Call `outputRates.config`. Then `lossRates` and `productionRates` will refer to the same species.
Fix the equations.
Fix headers in `mechanism.prod`?
Check stoichimetric coefficient handling.

**Comment by spco** [comment]
Output rate should be reaction rate divided by concentration of species of interest.

E.g.

`A+B--(k)-->C+D` has reaction rate `kAB`, so if `A` is species of interest, then `kB` is required.

Need to handle the cases `A+A -> B`, or `A->A+B`, `A->B+B` (and combinations).

**Comment by spco** [comment]
I don't like the way `reacDetailedRatesSpecies` and `prodDetailedRatesSpecies` contain the number of the species in the first column and then the numbers of the reactions in the following columns. This should be split up. The current name is also confusing, and should be applied only to the first column when split up.

**Comment by spco** [comment]
@rs028: a question on the desired outputs of lossRates and productionRates. Suppose I have the reaction `A+B ->C+D` with rate `k`. Then the reaction rate is `kAB`, where here and below I meanc 'the concentration of `A`' when i say `A`. I understand that if `A` is a species of interest then in `lossRates.output` there should be a line associated to it in this reaction, with value equal to `kB`. Similarly if `B` is of interest then we get `kAB`.

What happens if `C` is a species of interest? Should I expect a value in `productionRates.output` that is `kAB`? Or perhaps `kAB/C`? I just want to clarify, as that affects how loss rates are calculated versus production rates.

I'm ignoring the fact we want to later merge both these output files into one for now.

**Comment by rs028** [comment]
Given the reaction `A+B -> C+D`:

- species of interest is A
production of A is : NULL
loss of A is : kB

- species of interest is B
production of B is : NULL
loss of B is : kA

- species of interest is C
production of C is :  kAB
loss of C is : NULL

- species of interest is D
production of D is :  kAB
loss of D is : NULL

where I am using NULL to say that, for that reaction, species X is not produced (or not lost). Does it make sense?

**Comment by rs028** [comment]
Yes you are right. Keep in mind that those "reactions" are usually multiple reactions condensed into one for simplicity. What I was trying to do is to get the relative (to the species of interest) reaction rate, which can be  useful for some type of analysis.

However, I just found a note on the MCM website (here: http://mcm.leeds.ac.uk/MCM/tutorial_sec3.htt) that makes me think that the original code was meant to do something else, i.e. to reproduce a feature of the FACSIMILE software. Let me check and I will come back to you.

**Comment by rs028** [comment]
Okay, from the MCM website, the intended behaviour of the original code is  to reproduce a notation used by FACSIMILE:

    G25  % 1.90D-12*EXP(-1000/TEMP)         : OH + O3 = HO2            +S25;

where "The `G` values give the **instantaneous rate**, which is what we have defined in the above equations, and the `S` values the **cumulative rate** for each reaction."

As far as `G` is concerned, I am happy with what we do. The numbers in the `instantaneousRates` check out. So
```
A + B = C + D  --> kAB
A + A = B        --> kA^2
A = B + B        --> kA
```

However, `S` is not the "relative reaction rate" as I was thinking. I am trying to figure it out, the documentation is not clear on this, so I need to dig up some old models :)

**Comment by spco** [comment]
Ok. So to clarify my understanding - G is equivalent to our `instantaneousRates`, and that is all calculated correctly. S is the 'cumulative rate', but we're not quite sure what that means currently? I think you're right to hold off on any more of this until we understand better :)

**Comment by rs028** [comment]
That's right. I got my hand on some FACSIMILE model runs and I am going to figure out the `S`.

**Comment by rs028** [comment]
Ok, based on some simulations with the FACSIMILE software I think I understand what is going on.
Say we have a reaction `% k : A = B;`

- the instantaneous rates are calculated at a given time as `k*[A]*[B]`, which corresponds to the `G` parameter in FACSIMILE notation.

- `S` is the cumulative rate which at time T is calculated as `G * (number of timesteps since T-1) + (S at time T-1)`

So we have:

time    |  A      | k        | G       | S       |
-------|-------|-------|-------|-------|
43200 | 0              | 0             | 0               | 0              |
43500 | 4.88E+10 | 6.00E-06 |  2.93E+05 | 8.79E+07 = 2.93e5 * 300         |
43800 | 4.77E+10 | 6.00E-06 |  2.87E+05 | 1.74E+08 = 2.87e5 * 300 + 2.93e5 * 300  |

Does this make sense if we look at the current implementation? I can't understand from #318.

**Comment by spco** [comment]
Looks like `G` is corrrectly calculates as `instantaneousRates` in the code.

If we want `productionRates` to give `S` then there are problems.
`productionRates` is calculated as the cumulative sum of instantaneous rates per reaction, but with each reactant adding a contribution. I don't see any taking into account of the number of timesteps either. So, taking your example above but extending to `k : A = B + C`:

time    |  A      | k        | G       | S       |
-------|-------|-------|-------|-------|
43200 | 0              | 0             | 0               | 0              |
43500 | 4.88E+10 | 6.00E-06 |  2.93E+05 | 5.86e5 = 2.93e5 + 2.93e5         |
43800 | 4.77E+10 | 6.00E-06 |  2.87E+05 | 1.16e6 = 2.87e5 + 2.87e5 + 2.93e5 + 2.93e5  |

where we have 2 copies of each `G` being contributed because we have two reactants, each adding one copy. So we have something like `S at time T = (S at time T-1) + G * number of reactants in this reaction`.

**Comment by rs028** [comment]
That does not make much sense to me. What about lossRates? I think you said in #318 that it is calculated differently than productionRates?

**Comment by rs028** [comment]
At a second reading I am not sure I understand your example. `k : A = B + C` has two products not two reactants. did you want to say "product"?

**Comment by spco** [comment]
`lossRates` doesn't make much sense - for each of the species, it places the absolute value of the rate of loss of that species into the element of `lossRates` indexed by _reaction number_. This means that `lossRates` contains one number per reaction, and that number is just the rate of reaction of the _last product species in that reaction_. I can't see how that is possibly a useful number - it would give different results for `A = B + C` or `A = C + B`. Whatever it's doing, it doesn't seem sensible.

**Comment by spco** [comment]
For our future reference, the furthest commit back in the git history contains this in `resid()` in `solverFunction.f`:
```
    do i = 1,size1
    r(lhs(1,i)) = r(lhs(1,i))*y(lhs(2,i))**lhs(3,i)
    ir(lhs(1,i)) = r(lhs(1,i))
    enddo

    do i = 1,size1
        dy(lhs(2,i)) = dy(lhs(2,i))-lhs(3,i)*r(lhs(1,i))
        lossRates(lhs(1,i)) = abs(dy(lhs(2,i)))
    enddo

    do i = 1,size2
        dy(rhs(2,i))=dy(rhs(2,i))+coeff(i)*r(rhs(1,i))
        productionRates(rhs(1,i)) = productionRates(rhs(1,i)) + coeff(i)*r(rhs(1,i))
    enddo
```
which is exactly as we see it today, including the weirdness of the indexing of `lossRates`.

**Comment by rs028** [comment]
Okay, I think we can agree that the original code was bugged. My guess is that the implementation of this part of the code was not really finished.

It is not meant to calculate the relative reaction rates, as I initially thought. And I am not sure that a cumulative reaction rate is so useful that it needs to be standard output. My feeling is that it would be more useful to have a formatted version of instantaneous rates for selected species, taking into account the stoichiometry of the reaction. 

For example: if species of interest is **A**

reaction | instantaneousRates | lossRates
---------|--------|-------
A = B               | kA   | kA
A + A = B        | kAA | 2kAA
A + B = C        | kAB | kAB
A + B = A + C | kAB |kAB (equal to productionRate, so net = 0)

reaction | instantaneousRates | productionRates
---------|--------|-------
B = A               | kB   | kB
B = A + A        | kB   | 2kB
B + C = A        | kBC | kBC
A + B = A + C | kAB | kAB (equal to lossRate, so net = 0)

so that will be effectively multiply the instantaneous rates by the number of times the species of interest appears as a product or as a reactant (depending on what is being calculated). Does it make sense?

But first we need to check how the stoichimetric coefficient is handled by the python scripts that process the fac file or we end up multiplying for the same coefficient twice.

**Comment by rs028** [comment]
Also need to consider point 7 in issue #270 and first two points of issue #320.

**Comment by spco** [comment]
So far as I can see, the FACSIMILE format never contains a stoichiometric coefficient - it just relies on multiple copies of the same molecule - so you have to write `A+A` rather than `2A`. Googling doesn't turn anything up about this, nor does the MCM tutorial. Could you advise?

**Comment by spco** [comment]
http://mcm.leeds.ac.uk/MCM/tutorial_intro.htt doesn't seem to allow the use of stoichiometric coefficients in the FACSIMILE format, but I may be wrong.

Do we need to know whether the stoichiometric coefficient is >1? I don't see it breaking the loss/productionRates above, so I'm not sure why we'd need it?

**Comment by rs028** [comment]
Okay then :) You made a not at the beginning of this conversation to "Check stoichimetric coefficient handling." (in the python script).

That may not be necessary, but I want to be sure that the mechanism file is processed correctly (see my last comment https://github.com/AtChem/AtChem2/issues/319#issuecomment-416657197) and that the stoichiometric information is not lost.

That being said if the easiest way to calculate the production/loss rates is to just count the number of times a molecule appear as reactant/product in a reaction in the corresponding subroutine at runtime, I am fine with that. 

PS: I am in my office this afternoon if you prefer to talk this over.

**Comment by spco** [comment]
@rs028 , is it decided that we want to implement the production/loss rates as defined at https://github.com/AtChem/AtChem2/issues/319#issuecomment-416643702 ? If so, I probably have time to do this today or so.

**Comment by rs028** [comment]
Yes, that would be great!
I think that is the easiest way forward. We can always reimplement a cumulative rate calculation if needed be, but at the moment it doesn't seems very useful to me (besides the fact that it is wrong).

**Comment by spco** [comment]
Ok, and we want the format to be one line per reaction per species of interest per timestep? The move to a less 3D format can be done after the 1.1 milestone.

**Comment by rs028** [comment]
Yes, I think so. the format is a bit awkward to use but I would not consider that a priority. Let's focus on the 1.1 milestone.

**Comment by spco** [comment]
Closed by #351
